File: Java/src/main/java/com/gildedrose/AgedBrieItem.java
package com.gildedrose;

public class AgedBrieItem extends BaseItem {

    protected AgedBrieItem(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
    }

    @Override
    public void updateQuality() {
        if (this.quality < 50) {
            this.quality++;
        }
        if (this.sellIn < 0) {
            if (this.quality < 50) {
                this.quality++;
            }
        }
    }

}

File: scala/src/main/scala/com/gildedrose/AgedBrieItem.scala
package com.gildedrose

class AgedBrieItem(item: Item) extends BaseItem(item) {

  override def updateQuality: Unit = {
    increaseQuality
    if (item.sellIn < 0) {
      increaseQuality
    }
  }

  private def increaseQuality = {
    if (item.quality < 50) {
      item.quality += 1
    }
  }
}

object AgedBrieItem extends BaseFactory {

  def create(item: Item): Option[BaseItem] = if ("Aged Brie".equals(item.name)) Option(new AgedBrieItem(item)) else None;

}
